src/responder.ts

```diff
@@ -51,82 +62,85 @@ enum ResponderEvent {
         super(message);
         this.name = "NoNewBlockError";
     }
 }
 
+/**
+ * A simple custom Error class to signal that new blocks are being received, but the transaction
+ * is not being mined, suggesting it might be necessary to bump the gas price.
+ */
+export class StuckTransactionError extends Error {
+    constructor(message: string) {
+        super(message);
+        this.name = "StuckTransactionError";
+    }
+}
+
+
 /**
  * A generic abstract responder for the Ethereum blockchain.
+ * It has exclusive control of a wallet, that is, no two instances should share the same wallet.
  * It implements the submitStateFunction, but no strategy.
  */
 export abstract class EthereumResponder extends Responder {
     // TODO-93: the correct gas limit should be provided based on the appointment/integration.
     //          200000 is enough for Kitsune and Raiden (see https://github.com/raiden-network/raiden-contracts/blob/master/raiden_contracts/data/gas.json).
     private static GAS_LIMIT = 200000;
-    /**
-     * @param signer The signer of the wallet associated with this responder. Each responder should have exclusive access to his wallet.
-     * @param appointmentId The id of the Appointment this object is responding to.
-     * @param ethereumResponse The IEthereumResponse containing what needs to be submitted.
-     */
-    constructor(
-        readonly signer: ethers.Signer,
-        public readonly appointmentId: string,
-        public readonly ethereumResponse: IEthereumResponse
-    ) {
+
+    // implementations should query the provider (or a service) to figure out the appropriate gas price
+    protected gasPrice = new ethers.utils.BigNumber(21000000000);
+
+    // the nonce that will be used when submitStateFunction is called
+    protected nonce = null;
+
+    constructor(public readonly signer: ethers.Signer) {
         super();
     }
 
-    protected submitStateFunction(): Promise<TransactionResponse> {
+    protected submitStateFunction(responseData: IEthereumResponseData): Promise<TransactionResponse> {
         // form the interface so that we can serialise the args and the function name
-        const abiInterface = new ethers.utils.Interface(this.ethereumResponse.contractAbi);
-        const data = abiInterface.functions[this.ethereumResponse.functionName].encode(this.ethereumResponse.functionArgs);
+        const abiInterface = new ethers.utils.Interface(responseData.contractAbi);
+        const data = abiInterface.functions[responseData.functionName].encode(responseData.functionArgs);
         // now create a transaction, specifying possible oher variables
         const transactionRequest = {
-            to: this.ethereumResponse.contractAddress,
+            to: responseData.contractAddress,
             gasLimit: EthereumResponder.GAS_LIMIT,
-            // nonce: 0,
-            gasPrice: 21000000000, // TODO: choose an appropriate gas price
+            nonce: this.nonce,
```

Comment:
As above, ethers.js populates the nonce if left null or undefined (although the behaviour is not safe if, for example, many transactions are sent in bulk). I meant to leave that as default behaviour.

Reply:
Nota that the responder could in principle respond to many appointments, so nonces need to be managed properly. Actually, I think the best thing is removing the `nonce` class member, and pass it s a parameter of `submitStateFunction`, so I'll refactor it like that.

Reply:
Fair enough. It seems that you should remove line 223 then where you set the nonce = getTransactionCount(). Since at the moment you never use the null value.

Reply:
You've refactored to pass in a the nonce which is good. But you should remove the default as it is never used.

Reply:
Done in 03394fa08d6259fe4a1c44651a0d947fac0581df.